File: Scalvini VS Camuni/src/Arena.cpp
/*
 * Arena.cpp
 *
 *  Created on: 28 gen 2022
 *      Author: pietr
 */



#include <iostream>

#include "Arena.h"
#include "Camuno.h"
#include "Scalvino.h"
#include "ScalvinoCamuno.h"

//costruttore vuoto per l'arena
Arena::Arena() { }

//inizializzazione del singleton a NULL
Arena* Arena::arena = NULL;

//metodo di supporto per il singleton
//se l'istanza di arena non esiste, la crea e la restituisce
//altrimenti restituisce l'istanza già esistente
Arena* Arena::getArena() {
	if(arena == NULL)
		arena = new Arena();
	return arena;
}

//elimina dall'esistenza un concorrente sconfitto
void Arena::eliminaSconfitto(shared_ptr<Persona> p) { p.reset(); }

//gestisce i combattimenti fra persone basandosi sui loro punti forza
int Arena::combattimento(shared_ptr<Persona> p1, shared_ptr<Persona> p2){
	int idSconfitto;

	cout << "________________________________________________________" << endl << endl;

	//se si è scelto lo stesso personaggio
	if(p1->getID() == p2->getID()){
		cout << "--- " << p1->toString() << " non è ancora abbastanza ubriaco per combattere contro se stesso!" << endl;
		return -1;
	}else{
		//se i due personaggi hanno gli stessi punti forza
		if(p1->getForza() == p2->getForza()){
			cout << "--- " << p1->toString() << " e " << p2->toString() << " hanno pareggiato! -> Ancora in gioco!" << endl;
			return 0;
		}
		else{
			//p1 vincente, p2 perdente
			if(p1->getForza() > p2->getForza()){
				cout << "--- " << p1->toString() << " ha sconfitto " << p2->toString() << " ---" << endl;

				p1->modForza(2);		//la forza del vincente aumenta di 2 punti
				idSconfitto = p2->getID();
				eliminaSconfitto(p2);	//delete dello sconfitto
			}
			else	//p2 vincente, p1 perdente
			{
				cout << "--- " << p2->toString() << " ha sconfitto " << p1->toString() << " ---" << endl;

				p2->modForza(2);		//la forza del vincente aumenta di 2 punti
				idSconfitto = p1->getID();
				eliminaSconfitto(p1);	//delete dello sconfitto
			}
		}
	}

	return idSconfitto;	//restituisce l'id del personaggio sconfitto per la rimozione dalle liste
}


//distruttore
Arena::~Arena() { cout << "--- Arena smantellata ---" << endl << endl; }

File: Combattimenti Fantasy/src/Arena.scala
//arena di combattimento
object Arena {
  
  //combattimento tra due personaggi
  def combattimento(id1: Int, id2: Int): Unit = {
    
    var p1 = Gestore.getPersonaggio(id1)
    var p2 = Gestore.getPersonaggio(id2)
    
    //se almeno uno dei personaggi non esiste
    if(p1 == null || p2 == null){
      println("--- Combattimento non possibile, almeno uno dei due personaggi non esiste!")
      println
      return
    }
    
    //se i due personaggi coincidono
    if(p1.ID == p2.ID){
      println("--- " + p1.nick + " non è così ubriaco da combattere contro se stesso!")
      println("--- Ancora in gioco!")
      println
      return
    }
    
    //se vince il personaggio 1
    if(p1.forza > p2.forza){
      
      println("--- " + p1.nick + " ha sconfitto " + p2.nick + "!")
      println("--- La forza di " + p1.nick + " è incrementata di 5 punti!")
      
      Gestore.delPersonaggio(p2.ID)  //elimino il personaggio 2 dalla lista
      p1.forza_=(p1.forza + 5)       //incremento la forza del personaggio 1 di 5 punti
     
      return
    }
    
    //se vince il personaggio 2
    if(p2.forza > p1.forza){
     
      println("--- " + p2.nick + " ha sconfitto " + p1.nick + "!")
      println("--- La forza di " + p2.nick + " è incrementata di 5 punti!")
      
      Gestore.delPersonaggio(p1.ID)  //elimino il personaggio 1 dalla lista
      p2.forza_=(p2.forza + 5)       //incremento la forza del personaggio 2 di 5 punti
      
      return
    }
    
    //pareggio
    println("--- " + p1.nick + " e " + p2.nick + " hanno pareggiato!")
    println("--- Entrambi ancora in gioco!")
    println
     
  }
  
}